torch/nn/_functions/vision.py

```diff
@@ -93,35 +93,74 @@ def _enforce_cudnn(input):
-        if theta.is_cuda:
-            AffineGridGenerator._enforce_cudnn(theta)
-            assert False
-        ctx.is_cuda = False
-        base_grid = theta.new(N, H, W, 3)
-        linear_points = torch.linspace(-1, 1, W) if W > 1 else torch.Tensor([-1])
-        base_grid[:, :, :, 0] = torch.ger(torch.ones(H), linear_points).expand_as(base_grid[:, :, :, 0])
-        linear_points = torch.linspace(-1, 1, H) if H > 1 else torch.Tensor([-1])
-        base_grid[:, :, :, 1] = torch.ger(linear_points, torch.ones(W)).expand_as(base_grid[:, :, :, 1])
-        base_grid[:, :, :, 2] = 1
-        ctx.base_grid = base_grid
-        grid = torch.bmm(base_grid.view(N, H * W, 3), theta.transpose(1, 2))
-        grid = grid.view(N, H, W, 2)
+
+        if len(size) == 5:
+            N, C, D, H, W = size
+            ctx.size = size
+            ctx.is_cuda = theta.is_cuda
+            base_grid = theta.new(N, D, H, W, 4)
+
+            w_points = (torch.linspace(-1, 1, W) if W > 1 else torch.Tensor([-1]))
+            h_points = (torch.linspace(-1, 1, H) if H > 1 else torch.Tensor([-1])).unsqueeze(-1)
+            d_points = (torch.linspace(-1, 1, D) if D > 1 else torch.Tensor([-1])).unsqueeze(-1).unsqueeze(-1)
+
+            base_grid[:, :, :, :, 0] = w_points
+            base_grid[:, :, :, :, 1] = h_points
+            base_grid[:, :, :, :, 2] = d_points
+            base_grid[:, :, :, :, 3] = 1
+            ctx.base_grid = base_grid
+            grid = torch.bmm(base_grid.view(N, D * H * W, 4), theta.transpose(1, 2))
+            grid = grid.view(N, D, H, W, 3)
+
+        elif len(size) == 4:
+            N, C, H, W = size
+            ctx.size = size
+            if theta.is_cuda:
+                AffineGridGenerator._enforce_cudnn(theta)
+                assert False
+            ctx.is_cuda = False
+            base_grid = theta.new(N, H, W, 3)
+            linear_points = torch.linspace(-1, 1, W) if W > 1 else torch.Tensor([-1])
+            base_grid[:, :, :, 0] = torch.ger(torch.ones(H), linear_points).expand_as(base_grid[:, :, :, 0])
+            linear_points = torch.linspace(-1, 1, H) if H > 1 else torch.Tensor([-1])
+            base_grid[:, :, :, 1] = torch.ger(linear_points, torch.ones(W)).expand_as(base_grid[:, :, :, 1])
+            base_grid[:, :, :, 2] = 1
+            ctx.base_grid = base_grid
+            grid = torch.bmm(base_grid.view(N, H * W, 3), theta.transpose(1, 2))
+            grid = grid.view(N, H, W, 2)
+        else:
+            raise RuntimeError("AffineGridGenerator needs 4d (spatial) or 5d (volumetric) inputs.")
+
         return grid
 
     @staticmethod
     @once_differentiable
     def backward(ctx, grad_grid):
-        N, C, H, W = ctx.size
-        assert grad_grid.size() == torch.Size([N, H, W, 2])
-        assert ctx.is_cuda == grad_grid.is_cuda
-        if grad_grid.is_cuda:
-            AffineGridGenerator._enforce_cudnn(grad_grid)
+        if len(ctx.size) == 5:
+            N, C, D, H, W = ctx.size
```

Comment:
There seem to be a lot of redundant code between these two `if` cases. Can you make them share part of it?

Reply:
I'm happy to consolidate. I originally didn't to make it clear that the 2D case hadn't meaningfully changed.

Reply:
I see three lines in common:

        assert ctx.is_cuda == grad_grid.is_cuda
        base_grid = ctx.base_grid
        # ...
        grad_theta = grad_theta.transpose(1, 2)

I'll pull those out. I don't think that the other similarities are worth trying to de-duplicate, since the code would get significantly less readable.